### PR TITLE
Chore/fix encryption

### PR DESCRIPTION
**Pull request by YangXingHao96**
## Context

_Why does this PR exist? What problem are you trying to solve? What issue does this close?_
This PR is to address the issue of potential error in decryption of encrypted letterFields and issuedLetter. Previous decryption handles the cases of malformed utf8, empty results after decryption due to password error. However there is small chance that the text can be decrypted into utf8 context. Thus an anchorText is added to ensure that the message is correctly decrypted. For more details refer to the notion design doc: https://www.notion.so/opengov/20230719-Encryption-Decryption-Bug-Fix-0ff8890353a340f88dfc3e3f8a47a495

Closes [insert issue #]

## Approach

_How did you solve the problem?_

Additional column anchor text is added to letter table, perform database migration to add a field that is of type string and nullable, perform validation against default anchorText = OpenGovSg, which is saved as constant variable in repository. (should we modify this part to save it in database? like having an additional pre-encrypted anchor column?). If field is null, ignore and perform previous encryption, else, decrypt the content and validate against the default anchorText.

**Features**:

- Details ...

**Improvements**:

- Details ...

**Bug Fixes**:

- Details ...

## Before & After Screenshots

**BEFORE**:
[insert screenshot here]

**AFTER**:
[insert screenshot here]

## Tests

_What tests should be run to confirm functionality?_

## Deploy Notes

_Notes regarding deployment of the contained body of work. These should note any
new dependencies, new scripts, etc._

**New environment variables**:

- `env var` : env var details

**New scripts**:

- `script` : script details

**New dependencies**:

- `dependency` : dependency details

**New dev dependencies**:

- `dependency` : dependency details

## Risks

_Describes potential risks with this PR. Consider carefully what could possibly happen in a worse case scenario. If possible, try to define risks as 2 items: **Impact** and **Likelihood** and use your best judgment._

### CODE REVIEW

File: backend/src/database/entities/letter.entity.ts [hunk 1]
```diff
@@ -51,4 +51,7 @@ export class Letter {
 
   @Column({ type: 'timestamptz', nullable: true, default: null })
   firstReadAt: Date
+
+  @Column({ type: 'text', nullable: true })
+  anchorText: string
```
Comment (gweiying) [inline backend/src/database/entities/letter.entity.ts:56]: small suggestion, it was abit confusing to follow the anchorText variable throughout the code since we're reusing the variable name `anchorText` for both the encrypted anchor text and the decrypted anchor text. 

Let's call this column `encryptedAnchorText` for clarity?

File: backend/src/letters/letters-encryption.service.ts [hunk 2]
```diff
@@ -29,12 +32,21 @@ export class LettersEncryptionService {
       enc.Utf8,
     )
 
+    let anchorText: string
+    if (letter.anchorText === null) {
```
Comment (gweiying) [inline backend/src/letters/letters-encryption.service.ts:36]: Suggestion, let's add a short comment here to explain why anchorText might be null, something simple like `// we did not encrypt anchor text previously, so older password-protected letters do not contain the encryptedAnchorText column`

File: backend/src/letters/letters-encryption.service.ts [hunk 3]
```diff
@@ -29,12 +32,21 @@ export class LettersEncryptionService {
       enc.Utf8,
     )
 
+    let anchorText: string
```
Comment (gweiying) [inline backend/src/letters/letters-encryption.service.ts:35]: Suggestion, let's call this variable `decryptedAnchorText` for clarity, related to the comment above!

File: backend/src/database/migrations/1689908223807-add-anchortext.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { MigrationInterface, QueryRunner } from 'typeorm'
+
+export class addAnchortext1689908223807 implements MigrationInterface {
+  name = 'addAnchortext1689908223807'
```
Comment (gweiying) [inline backend/src/database/migrations/1689908223807-add-anchortext.ts:4]: tiny nit, mostly for future reference. would be nice to standardize on camelcase throughout the name, and also specify in the migration name which table the column is added to, for example `add-anchor-text-column-to-letters'

File: backend/src/letters/letters-encryption.service.spec.ts [hunk 5]
```diff
@@ -67,5 +85,25 @@ describe('LettersEncryptionService', () => {
         )
       }).toThrow() // wrongly decrypted long cypher texts can cause 'Malformed UTF-8 data'
     })
+
```
Comment (gweiying) [inline backend/src/letters/letters-encryption.service.spec.ts:88]: Small suggestion, what do you think of adding a test for backward compatibility, for when `anchorText` is null?

File: e2e/package.json [hunk 6]
```diff
@@ -10,6 +10,7 @@
     "playwright": "^1.34.3"
   },
   "devDependencies": {
-    "@playwright/test": "^1.34.3"
+    "@playwright/test": "^1.34.3",
```
Comment (gweiying) [inline e2e/package.json:13]: Nit, this and the `e2e/package-lock.json` changes are outside of the scope of this PR. Do you mind rebasing your branch off develop so we can keep the changes isolated?

**Review by gweiying — APPROVED** [review]
Overall the PR looks good! 

Left minor comments about naming clarity and tests. Your documentation in Notion was quite comprehensive, so thanks for that. I left a small suggestion for improvement there!

One other small thing:
Your branch contains changes from your `chore/fix_login_test` branch. To keep the changes isolated, could you grab the relevant changes and branch off `develop`? This would be the regular convention for our branching strategy, so good practice to do it now even though the changes are quite small here.

File: backend/src/letters/letters-encryption.service.ts (hunk 2 above)
Comment (YangXingHao96) [reply backend/src/letters/letters-encryption.service.ts:36]: added comments to indicate why it might be null

File: backend/src/letters/letters-encryption.service.ts (hunk 3 above)
Comment (YangXingHao96) [reply backend/src/letters/letters-encryption.service.ts:35]: changed the variable naming

File: backend/src/letters/letters-encryption.service.spec.ts (hunk 5 above)
Comment (YangXingHao96) [reply backend/src/letters/letters-encryption.service.spec.ts:88]: added a new test, that has anchorText as null

File: e2e/package.json (hunk 6 above)
Comment (YangXingHao96) [reply e2e/package.json:13]: removed all changes in e2e test
